Automattic/jetpack
# Search: Fix antipattern where components are returned from functions called renderXXXX() instead of just being components

Fixes unnecessary re-renders and lack of visibility in React Dev tools

### This PR is better reviewed seeing what each commit is doing

It's quite straightforward and if something here is wrong, the immediate effect is that the dashboard crashes

#### Changes proposed in this Pull Request:


* [x] Refactors renderFilterOption to be a component (Cherry-picked into #26426)
* [x] Refactors renderSearchResult to be a component (Cherry-picked into #26427)
* [x] Refactors renderInstantSearchToggle to be a component (Cherry-picked into #26428)
* [x] Refactors renderConnectScreen to be a component  (Cherry-picked into #26430)
* [x] Refactors renderConnectionFooter to be a component (cherry-picked into #26433)
* [x] Refactors renderHeader to be a component (Cherry-picked into #26435)
* [x] Refactors renderFooter to be a component  (Cherry-picked into #26435)
* [x] Refactors renderMockedSearchInterface to be a component (Cherry-picked into #26437)
* [x] Refactors renderModuleControl to just render component (Cherry-picked into #26437)

#### Other information:


#### Jetpack product discussion
None

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:
1. Launch a [JN site with Search standalone and this branch](https://jurassic.ninja/create/?jetpack-beta&branches.jetpack-search=fix/render-pattern-search&wp-debug-log)
2. Keep the JS console open and expect to see no warning for the following steps
3. Connect, navigate the page. Expect to see no crashing page
